File: ch7/js/c-subgraph-function.js
import { StateGraph, START, Annotation } from '@langchain/langgraph';

const StateAnnotation = Annotation.Root({
  foo: Annotation(),
});

const SubgraphStateAnnotation = Annotation.Root({
  // note that none of these keys are shared with the parent graph state
  bar: Annotation(),
  baz: Annotation(),
});

// Define subgraph
const subgraphNode = async (state) => {
  return { bar: state.bar + 'baz' };
};

const subgraph = new StateGraph(SubgraphStateAnnotation)
  .addNode('subgraph', subgraphNode)
  .addEdge(START, 'subgraph')
  // Additional subgraph setup would go here
  .compile();

// Define parent graph
const subgraphWrapperNode = async (state) => {
  // transform the state to the subgraph state
  const response = await subgraph.invoke({
    bar: state.foo,
  });
  // transform response back to the parent state
  return {
    foo: response.bar,
  };
};

const parentGraph = new StateGraph(StateAnnotation)
  .addNode('subgraph', subgraphWrapperNode)
  .addEdge(START, 'subgraph')
  // Additional parent graph setup would go here
  .compile();

// Example usage

const initialState = { foo: 'hello' };
const result = await parentGraph.invoke(initialState);
console.log(`Result: ${JSON.stringify(result)}`); // Should transform foo->bar, append "baz", then transform bar->foo

File: ch7/py/c-subgraph-function.py
from typing import TypedDict
from langgraph.graph import START, StateGraph


class State(TypedDict):
    foo: str


class SubgraphState(TypedDict):
    # none of these keys are shared with the parent graph state
    bar: str
    baz: str


# Define subgraph
def subgraph_node(state: SubgraphState):
    return {"bar": state["bar"] + "baz"}


subgraph_builder = StateGraph(SubgraphState)
subgraph_builder.add_node("subgraph_node", subgraph_node)
subgraph_builder.add_edge(START, "subgraph_node")
# Additional subgraph setup would go here
subgraph = subgraph_builder.compile()


# Define parent graph node that invokes subgraph
def node(state: State):
    # transform the state to the subgraph state
    response = subgraph.invoke({"bar": state["foo"]})
    # transform response back to the parent state
    return {"foo": response["bar"]}


builder = StateGraph(State)
# note that we are using `node` function instead of a compiled subgraph
builder.add_node("node", node)
builder.add_edge(START, "node")
# Additional parent graph setup would go here
graph = builder.compile()

# Example usage
initial_state = {"foo": "hello"}
result = graph.invoke(initial_state)
print(
    f"Result: {result}"
)  # Should transform foo->bar, append "baz", then transform bar->foo
